soroban-env-host/src/host.rs

```diff
@@ -2756,13 +2757,18 @@ impl VmCallerCheckedEnv for Host {
             for signer in signers {
                 if let SignerKey::Ed25519(ref this_signer) = signer.key {
                     if &target_signer == this_signer {
+                        // Clamp the weight at 255. Stellar protocol before v10
+                        // allowed weights to exceed 255, but the max threshold
+                        // is 255, hence there is no point in having a larger
+                        // weight.
+                        let weight = min(signer.weight, u8::MAX as u32);
                         // We've found the target signer in the account signers, so return the weight
-                        return Ok(signer.weight.into());
+                        return Ok(weight.into());
                     }
                 }
             }
-            // We didn't find the target signer, so it must have no weight
-            Ok(0u32.into())
+            // We didn't find the target signer, return an error
+            Err(self.err_status(ScHostObjErrorCode::ObjectNotExist))
```

Comment:
❗ Returning an error for this function is a problem, because it means for contracts that this function panics and a contract has no way of avoiding a panic. The pattern we've tried to follow as much as possible is that any panic should be avoidable either by:
- Making the host function return a `Status` when an error has occurred.
- Making the host function return some other value when a value makes sense, such as zero in this case.
- Adding a second host function that a contract can use to find out if calling this host function would panic, which in this case would mean adding a function like `account_has_signer`.

Given that in Stellar we say that a 0 weight is the same thing as not having that signer I think we should probably keep this as `0`. Soroban-auth, the token contract, and anyone else using this function can still see a `0` value and error in that case.

Reply:
I feel like this is yet another proof that we shouldn't be exposing this, as having such special values is pretty confusing. I also suspect that having 0 weight is not exactly the same as having no signer from the core perspective, but I could be missing some code that does the check. In any case I don't think we have any 0-weight signers in pubnet. So I guess it should be ok to return 0 for now.

Reply:
> as having such special values is pretty confusing

This special value exists on the Stellar network today. To remove a signer you set its weight to zero.

Reply:
Oh, I see, I must have missed that check.